pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
+          tableName, externalViewRecordVersion, lastKnownExternalViewVersion);
+      return true;
+    }
+
+    // Check if there are relevant instance config changes
+    Map<String, InstanceConfig> lastKnownInstanceConfigs = _lastKnownInstanceConfigsForTable.get(tableName);
+    if (lastKnownInstanceConfigs == null || lastKnownInstanceConfigs.isEmpty()) {
+      LOGGER.info("Routing table for table {} requires rebuild due to empty/null previous instance configs", tableName);
+      return true;
+    }
+
+    // Gather relevant incoming instance configs
+    Map<String, InstanceConfig> currentRelevantInstanceConfigs = new HashMap<>();
+    for (InstanceConfig incomingInstanceConfig : instanceConfigs) {
+      String instanceName = incomingInstanceConfig.getInstanceName();
+
+      if (lastKnownInstanceConfigs.containsKey(instanceName)) {
+        currentRelevantInstanceConfigs.put(instanceName, incomingInstanceConfig);
       }
+    }
 
+    // Did some instances lose their configuration?
+    if (lastKnownInstanceConfigs.size() != currentRelevantInstanceConfigs.size()) {
       LOGGER.info(
-          "Updating routing table for table {} due to ZK change (current version {}, last known version {})",
-          tableName, externalViewRecordVersion, lastKnownZkVersion);
+          "Routing table for table {} requires rebuild due to having a different number of instance configs (known instance config count {}, current instance config count {})",
+          tableName, lastKnownInstanceConfigs.size(), currentRelevantInstanceConfigs.size());
+      return true;
     }
 
-    _routingTableLastKnownZkVersionMap.put(tableName, externalViewRecordVersion);
+    // Did some instance change state?
+    for (String instanceName : lastKnownInstanceConfigs.keySet()) {
+      InstanceConfig previousInstanceConfig = lastKnownInstanceConfigs.get(instanceName);
+      InstanceConfig currentInstanceConfig = currentRelevantInstanceConfigs.get(instanceName);
+
+      // If it's the same znode, don't bother comparing the contents of the instance configs
+      if (previousInstanceConfig.getRecord().getVersion() == currentInstanceConfig.getRecord().getVersion()) {
+        continue;
+      }
+
+      // Check if the instance got enabled/disabled or started/stopped shutting down since the last update
+      boolean wasEnabled = previousInstanceConfig.getInstanceEnabled();
+      boolean isEnabled = currentInstanceConfig.getInstanceEnabled();
+
+      String wasShuttingDown =
+          previousInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+      String isShuttingDown =
+          currentInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+
+      boolean instancesChanged =
+          !EqualityUtils.isEqual(wasEnabled, isEnabled) || !EqualityUtils.isEqual(wasShuttingDown, isShuttingDown);
+
+      if (instancesChanged) {
```

Comment:
Am I right in that most of the time this will be 'false', because instanceConfig will change even if a segment is added to a server? If so, is there any way we can avoid notification for those changes? Not sure if there is a helix api for that.